### DISCUSSION TITLE
OpenAPI adoption for toolkit's major components

### DISCUSSION THREAD

**Discussion by luismojena** [fluxcd/flux2#828] [Proposals] [answered]
I would like to propose the adoption of OpenAPI standards for the major components in the toolkit. Aside from the fact that this is an industry-standard, this will also:

- Serve as documentation (can be automatically generated)
- Allow automatic generation of clients and server stubs in different languages
- Change the current specification pages (if you'll want that :wink: )
- Allow the community to better contribute 
- Allow the community to grow the ecosystem of libraries and applications

Let me know what you think!

**Comment by stefanprodan** [comment, marked as answer]
The GitOps Toolkit APIs are an extension to Kubernetes API expressed as [Custom Resource Definitions](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/). 

We use Kubernetes code generation tools for:

-  OpenAPI v3 schemas ([example](https://github.com/fluxcd/source-controller/blob/main/config/crd/bases/source.toolkit.fluxcd.io_gitrepositories.yaml#L34))
- API docs ([example](https://github.com/fluxcd/source-controller/blob/main/docs/api/source.md))
- Go API client ([example](https://github.com/fluxcd/source-controller/tree/main/api))

While we don't intend to provide any clients besides Golang, there are other projects in the Cloud-Native space like [Pulumi](https://www.pulumi.com/) and [AWS CDK8s](https://cdk8s.io) that allows you to interact with the toolkit API from various languages.
